pkg/agent/monitortool/monitor.go

```diff
@@ -0,0 +1,321 @@
+}
+
+func (n *NodeStore) Clear() {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	n.PingItems = make(map[string]PingItem)
+}
+
+func (n *NodeStore) AddPingItem(name string, ips []string) {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	n.PingItems[name] = PingItem{
+		Name: name,
+		IPs:  ips,
+	}
+}
+
+func (n *NodeStore) GetPingItem(name string) (PingItem, bool) {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+	item, found := n.PingItems[name]
+	return item, found
+}
+
+func (n *NodeStore) DeletePingItem(name string) {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	delete(n.PingItems, name)
+}
+
+func (n *NodeStore) ListPingItems() []PingItem {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+	items := make([]PingItem, 0, len(n.PingItems))
+	for _, item := range n.PingItems {
+		items = append(items, item)
+	}
+	return items
+}
+
+func (n *NodeStore) UpdatePingItem(name string, ips []string) {
+	n.mutex.Lock()
+	defer n.mutex.Unlock()
+	n.PingItems[name] = PingItem{
+		Name: name,
+		IPs:  ips,
+	}
+}
+
+func NewNodeLatencyMonitor(nodeInformer coreinformers.NodeInformer, interval, timeout time.Duration, limit int) *MonitorTool {
+	return &MonitorTool{
+		latencyStore: NewLatencyStore(nodeInformer),
+		interval:     interval,
+		timeout:      timeout,
+		limit:        limit,
+	}
+}
+
+func getRate(count int, d time.Duration, burst int) *rate.Limiter {
+	r := float64(count) / (d.Seconds() * 0.9)
+	return rate.NewLimiter(rate.Limit(r), burst)
+}
+
+func (m *MonitorTool) pingAll() {
+	// Get all node internal/external IP.
+	nodeIPs := m.latencyStore.ListNodeIPs()
+
+	// TODO: Get current node internal/external IP.
+	fromIP := ""
+
+	// A simple rate limiter
+	limiter := getRate(len(nodeIPs), m.interval, m.limit)
+	limitGroup := make(chan bool, int(m.limit))
+
+	klog.InfoS("Start to ping all nodes")
+	wg := sync.WaitGroup{}
+	for toIP, name := range nodeIPs {
+		limitGroup <- true
+		limiter.Wait(context.Background())
+		wg.Add(1)
+
+		go func(toIP string, name string) {
+			ctx, cancel := context.WithTimeout(context.Background(), m.timeout)
+			defer cancel()
+			defer wg.Done()
+
+			ok, rtt := m.pingNode(ctx, toIP)
+			if ok {
+				m.latencyStore.UpdateConnByKey(name, &Connection{
+					FromIP:      fromIP,
+					ToIP:        toIP,
+					Latency:     rtt,
+					Status:      true,
+					LastUpdated: time.Now(),
+				})
+			} else {
+				m.latencyStore.UpdateConnByKey(name, &Connection{
+					FromIP:      fromIP,
+					ToIP:        toIP,
+					Latency:     0,
+					Status:      false,
+					LastUpdated: time.Now(),
+				})
+			}
+
+			<-limitGroup
+
+		}(toIP, name)
+	}
+
+	wg.Wait()
+
+	// TODO: Print all connection status for debug
+	klog.InfoS("Finish to ping all nodes")
+	conns := m.latencyStore.ListConns()
+	for _, conn := range conns {
+		klog.InfoS("Connection status", "Connection", conn)
+	}
+}
+
+// resolveIP resolves the target address to an IP address.
+func resolveIPV4(ctx context.Context, target string) (*net.IPAddr, error) {
+	IPProtocol := "ip4"
```

Comment:
we should really have IPv6 support from the get go, I don't think it will add a lot of complexity

Reply:
Yes, I have updated it.